## Pulsar Classification For Class Prediction

- 17898 entries
- Data can be useful for prediction models of classification.

### COLUMNS:
Based on Integrated Profile of Observation
- Mean_Integrated: Mean of Observations
- SD: Standard deviation of Observations
- EK: Excess kurtosis of Observations
- Skewness: In probability theory and statistics, skewness is a measure of the asymmetry of the probability distribution of a real-valued random variable about its mean. Skewness of Observations.
- Mean _ DMSNR _ Curve: Mean of DM SNR CURVE of Observations
- SD _ DMSNR _ Curve: Standard deviation of DM SNR CURVE of Observations
- EK _ DMSNR _ Curve: Excess kurtosis of DM SNR CURVE of Observations
- Skewness _ DMSNR _ Curve: Skewness of DM SNR CURVE of Observations

### Class: Class 0 - 1

WHAT IS DM SNR CURVE:

- Radio waves emitted from pulsars reach earth after traveling long distances in space which is filled with free electrons.
- The important point is that pulsars emit a wide range of frequencies, and the amount by which the electrons slow down the wave depends on the frequency.
- Waves with higher frequency are sowed down less as compared to waves with lower frequency. It means dispersion.

TARGET:

Class
- 0 -- It is not
- 1 -- It is

## Import Libraries

In [332]:
import pandas as pd

## EDA

### 1. Understand the data

In [333]:
df=pd.read_csv('/content/PulsarDataMLProject.csv')

In [334]:
df.head()

Mean_Integrated         SD        EK  Skewness  Mean_DMSNR_Curve  \
0       140.562500  55.683782 -0.234571 -0.699648          3.199833   
1       102.507812  58.882430  0.465318 -0.515088          1.677258   
2       103.015625  39.341649  0.323328  1.051164          3.121237   
3       136.750000  57.178449 -0.068415 -0.636238          3.642977   
4        88.726562  40.672225  0.600866  1.123492          1.178930   

   SD_DMSNR_Curve  EK_DMSNR_Curve  Skewness_DMSNR_Curve  Class  
0       19.110426        7.975532             74.242225      0  
1       14.860146       10.576487            127.393580      0  
2       21.744669        7.735822             63.171909      0  
3       20.959280        6.896499             53.593661      0  
4       11.468720       14.269573            252.567306      0

In [335]:
df.shape

(17899, 9)

In [336]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17899 entries, 0 to 17898
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Mean_Integrated       17899 non-null  float64
 1   SD                    17898 non-null  float64
 2   EK                    17899 non-null  float64
 3   Skewness              17898 non-null  float64
 4   Mean_DMSNR_Curve      17899 non-null  float64
 5   SD_DMSNR_Curve        17896 non-null  float64
 6   EK_DMSNR_Curve        17899 non-null  float64
 7   Skewness_DMSNR_Curve  17899 non-null  float64
 8   Class                 17899 non-null  int64  
dtypes: float64(8), int64(1)
memory usage: 1.2 MB


In [337]:
df.describe()

Mean_Integrated            SD            EK      Skewness  \
count     17899.000000  17898.000000  17899.000000  17898.000000   
mean        111.075727     46.548516      0.478065      1.771610   
std          25.658495      6.844261      1.064374      6.170244   
min           5.812500     24.772042     -1.876011     -1.791886   
25%         100.925781     42.374622      0.027101     -0.188572   
50%         115.078125     46.946414      0.223241      0.198710   
75%         127.085938     51.023202      0.473406      0.928376   
max         192.617188     98.778911      8.069522     68.101622   

       Mean_DMSNR_Curve  SD_DMSNR_Curve  EK_DMSNR_Curve  Skewness_DMSNR_Curve  \
count      17899.000000    17896.000000    17899.000000          17899.000000   
mean          12.615600       26.330560        8.303198            104.851994   
std           29.472511       19.472907        4.506221            106.514308   
min            0.213211        7.370432       -3.139270             -1.976976   
25%            1.923077       14.437335        5.781447             34.949578   
50%            2.801839       18.465311        8.433157             83.060116   
75%            5.465719       28.432860       10.702946            139.307756   
max          223.392141      110.642211       34.539844           1191.000837   

              Class  
count  17899.000000  
mean       0.091625  
std        0.288504  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [338]:
df.duplicated().sum()

np.int64(1)

In [339]:
df.isnull().sum()

Mean_Integrated         0
SD                      1
EK                      0
Skewness                1
Mean_DMSNR_Curve        0
SD_DMSNR_Curve          3
EK_DMSNR_Curve          0
Skewness_DMSNR_Curve    0
Class                   0
dtype: int64

In [340]:
df['Class'].value_counts()

Class
0    16259
1     1640
Name: count, dtype: int64

#### Conclusion of EDA:

There is 1 duplicate record, 3 columns have null values. In total there are 9 features. We can use median to fill up the data into null spaces, duplicate record can be removed. There is imbalance in the Class with 16259 values in favour of "Not Pulsar" and 1639 in favour of "pulsar" class.


#### Choosing the right statistical model to classify the data
- Due to the large imbalance, we gotta choose "Random Forest" as it works well with imbalanced data.
- We need to check not only the accuracy but more importantly the precision, recall, f1 score as accuracy can be over 90% just by classifying record into 0 category.

### Data Cleaning

#### Drop duplicate record

In [341]:
df.drop_duplicates(inplace=True)

In [342]:
df.duplicated().sum()

np.int64(0)

#### Replace nulls with median of that feature

In [343]:
df['SD']=df['SD'].fillna(df['SD'].median())
df['Skewness']=df['Skewness'].fillna(df['Skewness'].median())
df['SD_DMSNR_Curve']=df['SD_DMSNR_Curve'].fillna(df['SD_DMSNR_Curve'].median())

In [344]:
df.isnull().sum()

Mean_Integrated         0
SD                      0
EK                      0
Skewness                0
Mean_DMSNR_Curve        0
SD_DMSNR_Curve          0
EK_DMSNR_Curve          0
Skewness_DMSNR_Curve    0
Class                   0
dtype: int64

### Analysis of feature correlation

In [345]:
df.corr()

Mean_Integrated        SD        EK  Skewness  \
Mean_Integrated              1.000000  0.547143 -0.873898 -0.738775   
SD                           0.547143  1.000000 -0.521438 -0.539791   
EK                          -0.873898 -0.521438  1.000000  0.945729   
Skewness                    -0.738775 -0.539791  0.945729  1.000000   
Mean_DMSNR_Curve            -0.298841  0.006873  0.414368  0.412057   
SD_DMSNR_Curve              -0.306997 -0.047613  0.432874  0.415137   
EK_DMSNR_Curve               0.234331  0.029421 -0.341209 -0.328844   
Skewness_DMSNR_Curve         0.144033  0.027680 -0.214491 -0.204781   
Class                       -0.673181 -0.363705  0.791591  0.709529   

                      Mean_DMSNR_Curve  SD_DMSNR_Curve  EK_DMSNR_Curve  \
Mean_Integrated              -0.298841       -0.306997        0.234331   
SD                            0.006873       -0.047613        0.029421   
EK                            0.414368        0.432874       -0.341209   
Skewness                      0.412057        0.415137       -0.328844   
Mean_DMSNR_Curve              1.000000        0.796557       -0.615971   
SD_DMSNR_Curve                0.796557        1.000000       -0.809783   
EK_DMSNR_Curve               -0.615971       -0.809783        1.000000   
Skewness_DMSNR_Curve         -0.354269       -0.575794        0.923743   
Class                         0.400876        0.491533       -0.390816   

                      Skewness_DMSNR_Curve     Class  
Mean_Integrated                   0.144033 -0.673181  
SD                                0.027680 -0.363705  
EK                               -0.214491  0.791591  
Skewness                         -0.204781  0.709529  
Mean_DMSNR_Curve                 -0.354269  0.400876  
SD_DMSNR_Curve                   -0.575794  0.491533  
EK_DMSNR_Curve                    0.923743 -0.390816  
Skewness_DMSNR_Curve              1.000000 -0.259117  
Class                            -0.259117  1.000000

#### Keeping:

- EK (best correlation with Class, 0.79)
- Skewness (second best correlation with Class, 0.70)
- Mean_Integrated (strong -0.67 correlation with Class)
- SD_DMSNR_Curve (moderate 0.49 correlation with Class)

#### Dropping:

- SD
- Skewness_DMSNR_Curve
- EK_DMSNR_Curve
- Mean_DMSNR_Curve


In [346]:
df.drop(['SD','Skewness_DMSNR_Curve','Mean_DMSNR_Curve','EK_DMSNR_Curve'],axis=1,inplace=True)

In [347]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17898 entries, 0 to 17898
Data columns (total 5 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Mean_Integrated  17898 non-null  float64
 1   EK               17898 non-null  float64
 2   Skewness         17898 non-null  float64
 3   SD_DMSNR_Curve   17898 non-null  float64
 4   Class            17898 non-null  int64  
dtypes: float64(4), int64(1)
memory usage: 839.0 KB


## Note:
Random Forest doesn't require scaling

## Train-test split

In [348]:
X=df.drop('Class',axis=1)
y=df['Class']

In [349]:
from sklearn.model_selection import train_test_split

In [350]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

## Model Building

In [356]:
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix

In [375]:
rf = RandomForestClassifier(n_estimators=500, max_depth=12, min_samples_leaf=10, class_weight={0:1, 1:3}, random_state=42, n_jobs=-1)

In [376]:
rf.fit(X_train, y_train)

RandomForestClassifier(class_weight={0: 1, 1: 3}, max_depth=12,
                       min_samples_leaf=10, n_estimators=500, n_jobs=-1,
                       random_state=42)

In [377]:
y_pred = rf.predict(X_test)

In [378]:
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification Report:\n", classification_report(y_test, y_pred, digits=3))

Confusion Matrix:
 [[3219   33]
 [  34  294]]

Classification Report:
               precision    recall  f1-score   support

           0      0.990     0.990     0.990      3252
           1      0.899     0.896     0.898       328

    accuracy                          0.981      3580
   macro avg      0.944     0.943     0.944      3580
weighted avg      0.981     0.981     0.981      3580

